File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/presentation/setting/AbstractPref.java
package com.lattis.ellipse.presentation.setting;

import android.content.SharedPreferences;

abstract class AbstractPref<T> {

    private SharedPreferences sharedPreferences;
    private String settingKey;
    private T defaultValue;

    AbstractPref(SharedPreferences sharedPreferences, String settingKey, T defaultValue) {
        this.sharedPreferences = sharedPreferences;
        this.settingKey = settingKey;
        this.defaultValue = defaultValue;
    }

    public abstract T getValue();

    T getDefaultValue(){
        return defaultValue;
    }

    public abstract void setValue(T value);

    SharedPreferences getSharedPreferences() {
        return sharedPreferences;
    }

    String getSettingKey() {
        return settingKey;
    }

    public void remove() {
        sharedPreferences.edit().remove(getSettingKey()).commit();
    }
}

File: velolabs_repos/lattis_android_v2/app/src/main/java/com/lattis/lattis/utils/settings/AbstractPref.kt
package com.lattis.lattis.utils.settings

import android.content.SharedPreferences

abstract class AbstractPref<T>(
    val sharedPreferences: SharedPreferences,
    val settingKey: String,
    val defaultValue: T
) {
    abstract var value: T

    fun remove() {
        sharedPreferences.edit().remove(settingKey).commit()
    }

}